### PR TITLE
Add repair of failed/expired rename in S3 backend

### PR DESCRIPTION
**Pull request by mosyp**
This PR brings the handling of stale/failed rename calls for S3 discussed in https://github.com/delta-io/delta-rs/discussions/89

### CODE REVIEW

File: rust/src/storage/s3/mod.rs
Comment (houqp) [inline rust/src/storage/s3/mod.rs:467]: Looks like there is no validation on lock data here, @mosyp just in case you missed the discussion in https://github.com/delta-io/delta-rs/discussions/89#discussioncomment-704821 :)

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:467]: Yes, I've definitely missed this one. That's a good catch, however I recall we specifically stated that the VersionsExists should be returned, let me have my check

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:467]: Git this, there was a case when it's required (more in discussion thread), but as it turns out that calling intended `rename` again would cover this

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:467]: @houqp @xianwill We have several choices here
Once the commit is repaired we
1) Release a lock, start acquire_loop again to get a lock with `data` we want, call rename, release lock (mostly reuse code)
2) Update a lock with new data, call rename, release lock (we have to create new update_lock function to update lock in dynamo (there's none, only put, get & release; and also we have to make a new rvn for a lock, so the lease_duration is extended, this is somewhat a _heartbeat_ from reference impl) 

We need to have updated data before calling the intended rename, otherwise, if it fails we'll most likely will break a log or lost it

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:467]: Also the problem with option 1 is that someone else might try to steal a lock and increase version etc, so it brings ambiguity to the situation

Comment (xianwill) [reply rust/src/storage/s3/mod.rs:467]: i like opt 2 - it seems much cleaner than opt 1 and the design you describe sounds robust and has a clear implementation path.

**Comment by mosyp** [comment]
FYI @rtyler @houqp @xianwill Tests to cover repair are on hold for PR until the end of a week as I want to dig into kafka-delta-ingest to catch up

File: rust/src/storage/s3/mod.rs
Comment (mosyp) [reply rust/src/storage/s3/mod.rs:495]: I'm not sure I follow you, we're not reacquire a lock here, e.g. we still hold it. The `update_data` just updates current lock's data, the `rvn` is updated though because we'd like to extends `lease_duration`, so that's why new lock instance is returned. 
This design is borrowed from reference implementation of `heartbeat`, on each change they also change a `rvn` so other workers would honour the changes and extends a wait

File: rust/src/storage/s3/mod.rs
Comment (mosyp) [reply rust/src/storage/s3/mod.rs:486]: hmm, that sounds legit, wdyt of ignoring a error from repair and just log it, or as you say, just recover `already_exists` only so it won't fail the intended rename?

Comment (houqp) [reply rust/src/storage/s3/mod.rs:486]: yes, in case of repair, the `already_exist` error should be ignored. other possible errors we should ignore are:

* copy failed due to `src` already deleted by concurrent repair from another writer
* delete failed due to `src` already deleted by concurrent repair from another writer

File: rust/src/storage/s3/mod.rs
Comment (houqp) [reply rust/src/storage/s3/mod.rs:493]: Yep, minor semantic difference. i consider a lock with new `rvn` a new lock in my previous comment. To me, the end result is not that different from releasing the lock, then reacquire the lock with the same data.

Let me try to add more context to the question I was asking with regards to `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS`.

Assuming no repair happened, the writer will try `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS` times in the acquire_lock_loop until it is able to acquire one. I think this semantic is very clear.

Now, let's consider the following trace:

* on the first try in `acquire_lock_loop`
* writer successfully acquired an expired lock. 
* writer completes the repair
* writer paused and lock expired
* writer resumed
* `self.update_data` failed updating lock with data

In this case, the lock update/reacquire error will be propagated back to the delta commit loop without a retry. Is this expected? In order words, do we want to make repair transparent to the caller so regardless repairs have happened or not, we always try to acquire the lock for `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS` times? If commit loop got a lock acquire error on the first try, what should it do?

File: rust/src/storage/s3/mod.rs
Comment (houqp) [reply rust/src/storage/s3/mod.rs:505]: cool, in this case, i think it might be better to return Ok with a state to signal different success cases. That way for callers who don't care about the details, they can just match on `Ok(_)`. Otherwise, I am a little bit concerned that some callers could introduce bugs by always assuming getting `Err(_)` means a rename failure.

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:505]: e.g. like a bool value? I saw that is a common practice for rename operations in some APIs

File: rust/src/storage/s3/mod.rs
Comment (mosyp) [reply rust/src/storage/s3/mod.rs:493]: To be honest, I'd like to crash a worker if it pauses, because why should one? However, if we see it's a common behaviour and with no workaround then we can make it resilient by allowing the worker to recover from this and retry rename. 

> do we want to make repair transparent to the caller so regardless repairs have happened or not, we always try to acquire the lock for DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS times? If commit loop got a lock acquire error on the first try, what should it do?

Crash I feel like, we should recover only on the error we assured, such as condition failure (someone else holds a lock) or throughput exceeded. These two are covered in reference impl as well. 

But regarding unexpected errors (and pauses as such) I lean toward failing the rename and worker itself, since most likely the owner should take actions

File: rust/src/storage/s3/mod.rs
Comment (mosyp) [reply rust/src/storage/s3/mod.rs:505]: Still, when thinking on this with perspective from https://github.com/delta-io/delta-rs/pull/233#discussion_r632431144 I feel like we might leave it as is here too, like yes, the rename succeeded but we still got unexpected error, which might involve to take some action, simply silent it and leave as is might lead into unexpected behaviour

**Comment by mosyp** [comment]
@xianwill @houqp @rtyler Recent commits are mainly a focused on rhttps://github.com/delta-io/delta-rs/pull/233/files#diff-477392fcbbeb53bddd769921d9b141e4cfa34728d1103a351fcaca2abd60f17f test which simulates worker pauses. 

There's 2 pauses, one before the rename is done (e.g. copy) and one is after - to cover two outputs when stale worker wakes up a) the copy will fail because the src is missing( already renamed) b) the releasing lock will fail, because it is expired

File: rust/src/storage/s3/dynamodb_lock.rs
Comment (xianwill) [inline rust/src/storage/s3/dynamodb_lock.rs:337]:
```suggestion
    /// The returned lock will have a new `rvn` so it'll increase the lease duration
```

File: rust/src/storage/s3/mod.rs
Comment (xianwill) [inline rust/src/storage/s3/mod.rs:479]:
```suggestion
    /// The returned lock will have a new `rvn` so it'll increase the lease duration
```

😆 - hoping my PR review contribution eventually amounts to more than doc edits

Comment (mosyp) [reply rust/src/storage/s3/mod.rs]: anyway that's a good ones, nobody likes to read some typos and get confused :)

File: rust/src/storage/s3/dynamodb_lock.rs
Comment (xianwill) [inline rust/src/storage/s3/dynamodb_lock.rs:27]: Do we need this env flag? Would it be possible to remove the added complexity and assume we will always delete data on release?

Comment (mosyp) [reply rust/src/storage/s3/dynamodb_lock.rs:27]: Yes, moreover, when thinking more about that since we don't need a data fields after the lock is released, we might just delete lock on release, as in one of the option in reference impl

**Review by xianwill — APPROVED** [review]

File: rust/src/storage/s3/mod.rs
Comment (houqp) [reply rust/src/storage/s3/mod.rs:524]: This is a fair design decision, but I think it's worth documenting somewhere in the doc string so users are aware that `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS` won't take into affect if lock acquisition failed due to worker itself pausing for too long.

File: rust/src/storage/s3/mod.rs
Comment (houqp) [reply rust/src/storage/s3/mod.rs:536]: In an API, typically this is done by returning Ok with a enum state to signal different cases. For example: `Ok(RenameResult::LockPreempted)`, `Ok(RenameResult::NoError)`, etc.

Then for applications that are OK with pauses, they can just match on `Ok(_)`. For applications that cannot live with pauses, they can explicitly match on `Ok(RenameResult::NoError)`:

```rust
match rename_with_lock() {
    Ok(RenameResult::NoError) => {},
    Ok(_) => {
        // Handle pauses 
    }
    Err(_) => // handle rename did not go through,
}
```

`Err` usually signals unrecoverable errors, so it's more likely that a client may assume this and check `Err(_)` for error, then proceed with retrying the rename with the same data even after the rename actually went through if we wrap this `lock is not released` error in `Err`.

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:536]: So at this point we assume that delta-rs write can live with pauses? E.g. wee still need to handle `RenameResult` into err or ok within try_commit loop

Comment (mosyp) [reply rust/src/storage/s3/mod.rs:536]: Btw, feels like we might expose an API as you described and leave it up to user on how to deal with it, through some config flag

File: rust/src/storage/s3/mod.rs
Comment (mosyp) [reply rust/src/storage/s3/mod.rs:524]: I agree that better documenting is needed, but I'm against to adapt to the case where network is slow and outreaches the default 20 seconds. For those cases we have an option to increase a `lease_duration` to whatever values matches their needs

Comment (houqp) [reply rust/src/storage/s3/mod.rs:536]: yeah, I wasn't suggesting we handle network pause here, I think you decision of crash loudly on pauses is a fine design decision. just need to be documented because since there are many distributed applications designed to be able to handle pauses, so not everyone will walk in with the same assumption :)

File: rust/src/storage/s3/mod.rs
Comment (houqp) [reply rust/src/storage/s3/mod.rs]: Yeah, I think delta-rs should be able to support use-cases that are ok with pauses and let the user decide whether they want to crash immediately on pause or continue as normal. Handling pauses automatically is a common design decision in distributed system designs.

**Comment by houqp** [comment]
@mosyp looks like there is minor merge conflicts that need to be addressed

**Comment by mosyp** [comment]
> @mosyp looks like there is minor merge conflicts that need to be addressed

Somebody been messing with dynamodb/s3 tests :)


Also, I'm going to revert your suggestion on `RenameResult` as of now and file a ticket for. I found that just introducing it to `rename` as `Ok` result is not enough because then optimistic concurrency will pick up a next version, which is not always desired. As we don't have an assurance that the actual version has been incremented by the worker that has stolen a lock, or any other one. So to avoid a gap between versions, we need to propagate an error of `LockPreempted` back to the optimistic concurrency loop so it'll retry with the same exact version OR get the latest one from `table.update()` call.

**Comment by mosyp** [comment]
Also, I'm going to revert your suggestion on `RenameResult` as of now and file a ticket for. I found that just introducing it to `rename` as `Ok` result is not enough because then optimistic concurrency will pick up a next version, which is not always desired. As we don't have an assurance that the actual version has been incremented by the worker that has stolen a lock, or any other one. So to avoid a gap between versions, we need to propagate an error of `LockPreempted` back to the optimistic concurrency loop so it'll retry with the same exact version OR get the latest one from `table.update()` call.

**Comment by houqp** [comment]
@mosyp perhaps I am missing something, when `Ok` is returned from the rename, it should indicate that the desired version has been committed regardless the lock has been expired or not. In case of `Ok(LockPreempted)`, we should notify the commit loop that the version has been successfully committed. `LockPreempted` only means the commit was first completed/repaired by another writer right? Why do we need to retry with the exact same version?

**Comment by mosyp** [comment]
@houqp Yes, my assumption was incorrect, however I'm glad we'll cover that in separate topic since this PR was already overwhelmed

**Pull request merged**
